collector/cpu_linux.go

```diff
@@ -203,7 +206,11 @@ func (c *cpuCollector) updateStat(ch chan<- prometheus.Metric) error {
 		return err
 	}
 
-	for cpuID, cpuStat := range stats.CPU {
+	c.updateCPUStats(stats.CPU)
+
+	// Acquire a lock to read the stats.
+	c.cpuStatsMutex.Lock()
```

Comment:
can we take the lock for this whole function instead?

Reply:
I wanted to keep the main update lock inside the updateCPUStats function. This keeps the lock local to where it needs to be for updating the data.

Reply:
From an execution perspective, if two scrapes are updating the stats at the same time, one will lock, then the other, then we would enter in to the read phase. By releasing and re-acquiring the lock, we get the last updated write.

Reply:
OK, what about defering the unlock?

As we use "MustNewConstMetric" functions it means that they risk `panic`-ing (they should not but defering the unlock would make us more resilient).

Reply:
> From an execution perspective, if two scrapes are updating the stats at the same time, one will lock, then the other, then we would enter in to the read phase. By releasing and re-acquiring the lock, we get the last updated write.

Indeed it makes sense

Reply:
If they panic, wouldn't that crash the whole exporter? Maybe not?

Reply:
Updated to defer.

Reply:
No the http handler would recover it